### PR TITLE
htlcswitch+routerrpc: support unencrypted failure reasons

### PR DESCRIPTION
**Pull request by joostjager**
This PR adds the possibility for users of the htlc interception api to supply arbitrary pre-encoded failure messages. LND will only add an hmac and encrypt the message for the first hop.

Interceptor applications that implement a virtual channel can use this to return for example a `FeeInsufficient` failure as if it was generated within the LND node.

### CODE REVIEW

File: htlcswitch/interceptable_switch.go [hunk 1]
```diff
@@ -119,10 +119,15 @@ type FwdResolution struct {
...
-	// FailureMessage is the encrypted failure message that is to be passed
-	// back to the sender if action is FwdActionFail.
+	// FailureMessage is the failure message that is to be passed back to
+	// the sender if action is FwdActionFail.
 	FailureMessage []byte
 
+	// FailureMessageEncrypted indicates whether FailureMessage is already
+	// encrypted for the first hop and just needs to be intermediate-node
+	// obfuscated.
+	FailureMessageEncrypted bool
```
Comment (champo) [inline htlcswitch/interceptable_switch.go:129]: should this rather be "is last hop" or similar?

File: htlcswitch/interceptable_switch.go [hunk 2]
```diff
@@ -378,7 +383,10 @@ func (s *InterceptableSwitch) resolve(res *FwdResolution) error {
 
 	case FwdActionFail:
 		if len(res.FailureMessage) > 0 {
-			return intercepted.Fail(res.FailureMessage)
+			return intercepted.Fail(
+				res.FailureMessage,
+				!res.FailureMessageEncrypted,
```
Comment (champo) [inline htlcswitch/interceptable_switch.go:388]: shouldn't this keep the same semantics? why change them here?

File: lnrpc/routerrpc/forward_interceptor.go [hunk 3]
```diff
@@ -168,12 +171,11 @@ func (r *forwardInterceptor) resolveFromClient(
...
-	resolution *htlcswitch.FwdResolution) error {
+	encrypted bool, resolution *htlcswitch.FwdResolution) error {
 
-	// Verify that the size is equal to the fixed failure
-	// message size + hmac + two uint16 lengths. See BOLT
-	// #4.
-	if len(failMessage) !=
+	// Verify that for encrypted messages the size is equal to the fixed
+	// failure message size + hmac + two uint16 lengths. See BOLT #4.
+	if encrypted && len(failMessage) !=
```
Comment (champo) [inline lnrpc/routerrpc/forward_interceptor.go:178]: would it be reasonable to try and decode the message here if it's not encrypted? it would be the "equivalent" check

File: htlcswitch/interceptable_switch.go (hunk 2 above)
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:388]: I thought "encrypted" is the normal form, so wanted to use that as much as possible. But to keep the rpc backwards compatible, I needed `failuremessage_unencrypted` which defaults to false. Therefore the conversion.

File: lnrpc/routerrpc/forward_interceptor.go (hunk 3 above)
Comment (joostjager) [reply lnrpc/routerrpc/forward_interceptor.go:178]: What would be the purpose to try to decode? If you mean to avoid the new flag, I think that might be risky in case an encrypted failure message accidentally decodes.

Comment (champo) [reply lnrpc/routerrpc/forward_interceptor.go:178]: I didn't mean to avoid the flag! mostly a sanity check that the message is valid. if the interceptor is returning non-valid messages, it can be quite tricky to debug. I meant something like this:

```
if encrypted {
  // len check
} else {
 // decode check
}}
```

File: htlcswitch/interceptable_switch.go (hunk 2 above)
Comment (champo) [reply htlcswitch/interceptable_switch.go:388]: oh, didn't think about the backwards compat of the field. makes sense!

File: htlcswitch/interceptable_switch.go (hunk 1 above)
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:129]: I'd rather keep it more low-level and only describe the contents of the failure message field.

File: lnrpc/routerrpc/forward_interceptor.go (hunk 3 above)
Comment (joostjager) [reply lnrpc/routerrpc/forward_interceptor.go]: Great suggestion, added

**Review by champo — APPROVED** [review]

**Comment by alexbosworth** [comment]
not that i think this is in scope of the issue but it would also be cool generally if you could return the error as if it was from the next hop, if you know the private key associated with the next hop like for a virtual peer returning an invalid payment hash error

**Comment by joostjager** [comment]
Assuming I understand the comment correctly, this is already supported on the htlc interceptor api via an encrypted failure message? For example in lnmux, we [encrypt for the sender](https://github.com/bottlepay/lnmux/blob/e5c4a4cadd4ca9a1aa8967bbbe51d336b878d466/mux.go#L298) outside of lnd.

**Comment by alexbosworth** [comment]
> Assuming I understand the comment correctly, this is already supported on the htlc interceptor api via an encrypted failure message? For example in lnmux, we [encrypt for the sender](https://github.com/bottlepay/lnmux/blob/e5c4a4cadd4ca9a1aa8967bbbe51d336b878d466/mux.go#L298) outside of lnd.

yes exactly, what i mean is to do this at the API level in an unencrypted way, to avoid the caller having to build the encrypted failure message

File: lnrpc/routerrpc/router.proto [hunk 4]
```diff
@@ -818,12 +818,18 @@ message ForwardHtlcInterceptResponse {
...
     //
-    // If failure_message is specified, the failure_code field must be set
-    // to zero.
+    // If failure_message is specified, the failure_code field must be set to
+    // zero.
     bytes failure_message = 4;
 
+    // Indicates whether the failure message still needs to be encrypted for the
+    // first hop.
+    bool failure_message_unencrypted = 6;
```
Comment (joostjager) [inline lnrpc/routerrpc/router.proto:831]: >yes exactly, what i mean is to do this at the API level in an unencrypted way, to avoid the caller having to build the encrypted failure message

Okay got it. I think it is debatable whether this is functionality that should live inside lnd. Passing in the private key is a security risk too.

File: lnrpc/routerrpc/forward_interceptor.go [hunk 5]
```diff
@@ -125,52 +131,18 @@ func (r *forwardInterceptor) resolveFromClient(
...
-			// message size + hmac + two uint16 lengths. See BOLT
-			// #4.
-			if len(in.FailureMessage) !=
-				lnwire.FailureMessageLength+32+2+2 {
-
-				return status.Errorf(
-					codes.InvalidArgument,
-					"failure message length invalid",
-				)
+			err := r.handleFailMessage(in.FailureMessage, fwdRes)
```
Comment (halseth) [inline lnrpc/routerrpc/forward_interceptor.go]: clearer pattern IMO:
```
fwdRes.FailMessage, err = r.handleFailMessage(in.FailureMessage)
```

File: lnrpc/routerrpc/forward_interceptor.go [hunk 6]
```diff
@@ -194,3 +166,54 @@ func (r *forwardInterceptor) resolveFromClient(
 		)
 	}
 }
+
+func (r *forwardInterceptor) handleFailMessage(failMessage []byte,
```
Comment (halseth) [inline lnrpc/routerrpc/forward_interceptor.go:174]: does this function have to be on the `forwardInterceptor`?

File: lnrpc/routerrpc/forward_interceptor.go (hunk 5 above)
Comment (joostjager) [reply lnrpc/routerrpc/forward_interceptor.go]: in a later commit, another field is set on `fwdRes`.

File: lnrpc/routerrpc/forward_interceptor.go (hunk 6 above)
Comment (joostjager) [reply lnrpc/routerrpc/forward_interceptor.go]: good point, moved validation to interceptable switch. it goes deeper than just a bit of input checking.

File: lnrpc/routerrpc/forward_interceptor.go [hunk 7]
```diff
@@ -125,52 +130,28 @@ func (r *forwardInterceptor) resolveFromClient(
...
-			code = lnwire.CodeInvalidOnionHmac
-
-		case lnrpc.Failure_INVALID_ONION_KEY:
-			code = lnwire.CodeInvalidOnionKey
-
-		case lnrpc.Failure_INVALID_ONION_VERSION:
-			code = lnwire.CodeInvalidOnionVersion
+		err := r.htlcSwitch.Resolve(fwdRes)
+		if errors.Is(err, htlcswitch.ErrInvalidFailureMsgLength) ||
+			errors.Is(err, htlcswitch.ErrUnparseableFailureMsg) {
```
Comment (joostjager) [inline lnrpc/routerrpc/forward_interceptor.go:147]: Caused by `errorlint` where this could have been a clean and familiar switch statement. Urging reconsideration.

File: lnrpc/routerrpc/forward_interceptor.go (hunk 5 above)
Comment (joostjager) [reply lnrpc/routerrpc/forward_interceptor.go]: Comment no longer applies, validation moved to `InterceptableSwitch`.

File: lnrpc/routerrpc/forward_interceptor.go (hunk 7 above)
Comment (joostjager) [reply lnrpc/routerrpc/forward_interceptor.go:147]: Yes. Also trying to think of problems that we had in the past with not checking the whole chain. I don't think it really happened, or rarely.

In my opinion, it's not like for example binding the loop variable, which happened a lot and has serious consequences.

Comment (halseth) [reply lnrpc/routerrpc/forward_interceptor.go:147]: what about 
```
switch {
case errors.Is(err, <err1>):
        fallthrough
case errors.Is(err, <err2>):
        return status.Errorf(...)
}
```

**Review by halseth — APPROVED** [review]
LGTM 💯

File: lnrpc/routerrpc/forward_interceptor.go (hunk 7 above)
Comment (joostjager) [reply lnrpc/routerrpc/forward_interceptor.go:147]: Don't like it. I just don't think the `errors.Is` is helping us much. Almost like JavaScript, use == on any runtime determined type and it just works.

File: htlcswitch/interceptable_switch.go [hunk 8]
```diff
@@ -390,6 +408,23 @@ func (s *InterceptableSwitch) resolve(res *FwdResolution) error {
...
+
+		return ErrInvalidFailureMsgLength
+
+	// For unencrypted messages, verify that they are parseable.
+	case !res.FailureMessageEncrypted:
+		r := bytes.NewReader(res.FailureMessage)
+		_, err := lnwire.DecodeFailureMessage(r, 0)
+		if err != nil {
+			return ErrUnparseableFailureMsg
+		}
```
Comment (joostjager) [inline htlcswitch/interceptable_switch.go:425]: I want to add extra code here to re-sign a channel update with this node's key if it is part of the decoded failure message.

Comment (joostjager) [reply htlcswitch/interceptable_switch.go:484]: Tacked on commit for this.

**Comment by joostjager** [comment]
!lightninglabs-deploy mute

**Comment by joostjager** [comment]
Rebased

File: lnrpc/routerrpc/router.proto (hunk 4 above)
Comment (Roasbeef) [reply lnrpc/routerrpc/router.proto:927]: This discussion thread still relevant?

File: htlcswitch/interceptable_switch.go [hunk 9]
```diff
@@ -93,6 +94,11 @@ type InterceptableSwitch struct {
 	// currentHeight is the currently best known height.
 	currentHeight int32
 
+	// signChannelUpdate is used when an intercepting application includes
+	// an unsigned channel update to be signed by us.
+	signChannelUpdate func(u *lnwire.ChannelUpdate) (*ecdsa.Signature,
```
Comment (Roasbeef) [inline htlcswitch/interceptable_switch.go:99]: Hmm, shouldn't we assume that they want to instead send the latest update? Otherwise i can see this causing an error where: user sends custom channel update, the one on disk conflicts, so routing errors occur.

File: htlcswitch/interceptable_switch.go [hunk 10]
```diff
@@ -173,6 +179,11 @@ type InterceptableSwitchConfig struct {
 	// RequireInterceptor indicates whether processing should block if no
 	// interceptor is connected.
 	RequireInterceptor bool
+
+	// SignAliasUpdate is used when an intercepting application includes an
+	// unsigned channel update to be signed by us.
+	SignChannelUpdate func(u *lnwire.ChannelUpdate) (*ecdsa.Signature,
```
Comment (Roasbeef) [inline htlcswitch/interceptable_switch.go:185]: Godoc mismatch, but see above, not sure we want to allow this.

File: htlcswitch/interceptable_switch.go (hunk 9 above)
Comment (Roasbeef) [reply htlcswitch/interceptable_switch.go:99]: We can prevent this in the forwarding level during the validation of the returned message. If it's an encoded channel update, then it needs to match what we'd send out anyway.

File: htlcswitch/interceptable_switch.go [hunk 11]
```diff
@@ -406,6 +418,49 @@ func (s *InterceptableSwitch) resolve(res *FwdResolution) error {
...
+	sig, err := s.signChannelUpdate(update)
+	if err != nil {
+		return err
+	}
+
+	update.Signature, err = lnwire.NewSigFromSignature(sig)
+	if err != nil {
+		return nil
+	}
+
```
Comment (Roasbeef) [inline htlcswitch/interceptable_switch.go:434]: Doesn't commit it. Even if we do decide to allow this, we also need to make sure that the update would actually validate in the first place (has all the needed fields, bounds respected, etc).

File: htlcswitch/interceptable_switch.go [hunk 12]
```diff
@@ -428,10 +483,30 @@ func (s *InterceptableSwitch) validateResolution(res *FwdResolution) error {
...
+			// Re-encode the failure message for delivery to the
+			// sender.
+			var w bytes.Buffer
+
+			err = lnwire.EncodeFailureMessage(&w, msg, 0)
+			if err != nil {
+				return err
+			}
+
+			res.FailureMessage = w.Bytes()
```
Comment (Roasbeef) [inline htlcswitch/interceptable_switch.go:508]: IMO this shouldn't mutate the failure message from underneath the caller. This method currently handles just validation, so should keep doing that one thing. 

If we still like this re-sign portion, then we should move it to a more specific layer.

**Review by Roasbeef — CHANGES_REQUESTED** [review]

File: htlcswitch/interceptable_switch.go [hunk 13]
```diff
@@ -388,8 +393,34 @@ func (s *InterceptableSwitch) resolve(res *FwdResolution) error {
...
+
+	// No validation if there is no failure message.
+	if res.FailureMessage == nil {
+		return nil
+	}
+
+	// Verify that the size is equal to the fixed failure
+	// message size + hmac + two uint16 lengths. See BOLT
+	// #4.
+	if len(res.FailureMessage) != lnwire.FailureMessageLength+32+2+2 {
```
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/interceptable_switch.go]: Added commit that replaces `32` with `sha256.Size`. But not sure how you'd remove the remaining magic?

File: htlcswitch/interceptable_switch.go [hunk 14]
```diff
@@ -646,10 +669,25 @@ func (f *interceptedForward) Resume() error {
...
+		obfuscatedReason []byte
+		obfuscator       = f.packet.obfuscator
+	)
+
+	if encryptFirstHop {
+		var err error
+		obfuscatedReason, err = obfuscator.EncryptEncodedFirstHop(
+			reason,
+		)
+
```
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:759]: Removed

File: lnrpc/routerrpc/router.proto (hunk 4 above)
Comment (joostjager) [reply lnrpc/routerrpc/router.proto:927]: It is a good point, and indeed just passing in the shared secret for the virtual hop and letting lnd do the encryption could be a middle way.

This PR is about adding support for unencrypted failure reasons though, not making it easier to use the encrypted reason functionality that is pre-existing. I think that would be more suitable as a follow-up? In that a `failure_message_shared_secret` could be added to indicate that `failure_message` still has to be encrypted. At that point, `failure_message_unencrypted = false` can be deprecated.

File: lnrpc/routerrpc/router.proto [hunk 15]
```diff
@@ -914,20 +914,28 @@ message ForwardHtlcInterceptResponse {
...
+    //
     // Return the specified failure code in case the resolve action is Fail. The
     // message data fields are populated automatically.
     //
     // If a non-zero failure_code is specified, failure_message must not be set.
     //
     // For backwards-compatibility reasons, TEMPORARY_CHANNEL_FAILURE is the
     // default value for this field.
-    lnrpc.Failure.FailureCode failure_code = 5;
+    lnrpc.Failure.FailureCode failure_code = 5 [deprecated = true];
```
[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/routerrpc/router.proto:938]: The thought here was that passing just a code isn't a very complete API, and once you start adding additional failure message fields you can just as well pass in a complete encoded, unencrypted failure message.

But perhaps there are use cases where you'd explicitly want lnd to fill in those missing fields.

Removed deprecated marker for now.

File: htlcswitch/interfaces.go [hunk 16]
```diff
@@ -339,7 +339,7 @@ type InterceptedForward interface {
 
 	// Fail notifies the intention to fail an existing hold forward with an
 	// encrypted failure reason.
-	Fail(reason []byte) error
+	Fail(reason []byte, encryptFirstHop bool) error
```
[the remark this answers is not in the document]
Comment (joostjager) [reply htlcswitch/interfaces.go:342]: Updated

File: htlcswitch/interceptable_switch.go (hunk 9 above)
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:99]: The context in which this is to be used is to update the routing policy of a virtual channel. The channel isn't broadcasted through gossip and doesn't exist in the graph on disk. There is no 'latest update'. 

The existence of the virtual channel is solely communicated to the payer through invoice route hints. If the receiver (the virtual hop) changes its policy, the channel update contained in the failure message updates the sender with the new values.

File: htlcswitch/interceptable_switch.go (hunk 10 above)
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:185]: Fixed godoc. Discussion can continue in https://github.com/lightningnetwork/lnd/pull/7067#discussion_r1201447182

File: htlcswitch/interceptable_switch.go (hunk 12 above)
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:508]: Agreed that this is an obscure side effect.

Separated validation and resigning. It is still mutating the resolution to avoid copying the failure message for which there doesn't seem to exist a standard method. But it is much more explicit now.

File: htlcswitch/interceptable_switch.go (hunk 11 above)
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:434]: You mean commit as in write to disk? As explained in https://github.com/lightningnetwork/lnd/pull/7067/commits/e4af8d7a4856d0d6597a264aed74f0df8834a749#r1202002584, it is the intention to just pass through without persisting.

File: lnrpc/routerrpc/forward_interceptor.go [hunk 17]
```diff
@@ -130,35 +135,25 @@ func (r *forwardInterceptor) resolveFromClient(
...
-				lnwire.FailureMessageLength+32+2+2 {
-
-				return status.Errorf(
-					codes.InvalidArgument,
-					"failure message length invalid",
-				)
+			// Fail with an encrypted reason.
+			fwdRes.FailureMessage = in.FailureMessage
+		} else {
+			code, err := unmarshalFailCode(in.FailureCode)
```
[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/routerrpc/forward_interceptor.go:142]: Created `unmarshallFailureResolution` function in the rpc server that also validates.

File: lnrpc/routerrpc/forward_interceptor.go [hunk 18]
```diff
@@ -121,45 +121,18 @@ func (r *forwardInterceptor) resolveFromClient(
...
-				FailureMessage: in.FailureMessage,
-			})
+		// Instantiate the fwdRes with base fields set.
+		fwdRes := &htlcswitch.FwdResolution{
+			Key:    circuitKey,
+			Action: htlcswitch.FwdActionFail,
 		}
 
-		code, err := unmarshalFailCode(in.FailureCode)
+		err := unmarshallFailureResolution(in, fwdRes)
```
Comment (Roasbeef) [inline lnrpc/routerrpc/forward_interceptor.go:131]: Don't think it's blocking, but we could have this instead return the new value and assign it here, s.t the new function is a pure function. Then tests can just exercise that the right value is returned vs asserting that the input is mutated as expected.

File: lnrpc/routerrpc/router.proto (hunk 4 above)
Comment (Roasbeef) [reply lnrpc/routerrpc/router.proto:927]: @joostjager IIUC, this is intended for a use case for "virtual nodes", so anycast-like behavior using special hop hints. In that case, how does this help? It'll let you send a _plaintext_ error (from this interceptor sub-system) and have lnd encrypted that, where it's the source node. IIUC though, you actually want to be able to send errors from the "virtual node", which means you need to use the shared secret for _that_ hop for the path finding retry/apply logic to work properly.

Comment (Roasbeef) [reply lnrpc/routerrpc/router.proto:927]: @joostjager IIUC, this is intended for a use case for "virtual nodes", so anycast-like behavior using special hop hints. In that case, how does this help? It'll let you send a _plaintext_ error (from this interceptor sub-system) and have lnd encrypted that, where it's the source node. IIUC though, you actually want to be able to send errors from the "virtual node", which means you need to use the shared secret for _that_ hop for the path finding retry/apply logic to work properly.

File: htlcswitch/interceptable_switch.go (hunk 9 above)
Comment (Roasbeef) [reply htlcswitch/interceptable_switch.go:90]: > The context in which this is to be used is to update the routing policy of a virtual channel. The channel isn't broadcasted through gossip and doesn't exist in the graph on disk.

Assuming the virtual channel has a distinct pubkey (and also shared secret in the onion), then how would this work in practice? lnd would use diff logic for if it was the final hop vs the penultimate hop.

File: htlcswitch/interceptable_switch.go [hunk 19]
```diff
@@ -394,9 +406,26 @@ func (s *InterceptableSwitch) resolve(res *FwdResolution) error {
...
+			// Re-sign the channel update if present. Note that this
+			// changes the passed in FwdResolution.
+			update := getChannelUpdateRef(msg)
+			if update != nil {
+				err := s.validateChannelUpdate(update)
+				if err != nil {
+					return err
+				}
+
+				err = s.resignChannelUpdate(update)
```
Comment (Roasbeef) [inline htlcswitch/interceptable_switch.go:420]: Should we also commit this update to disk as well? Apologies if this was already an older review comment...

File: htlcswitch/interceptable_switch.go [hunk 20]
```diff
@@ -416,6 +445,60 @@ func (s *InterceptableSwitch) resolve(res *FwdResolution) error {
 	}
 }
 
+func (s *InterceptableSwitch) validateChannelUpdate(
+	update *lnwire.ChannelUpdate) error {
+
+	// The maxHTLC flag is mandatory.
+	if !update.MessageFlags.HasMaxHtlc() {
```
Comment (Roasbeef) [inline htlcswitch/interceptable_switch.go:452]: More validation is needed here, eg: max HTLC shouldn't be greater than the channel capacity, etc, etc. I think we should just call the normal validation routine.

Comment (Roasbeef) [inline htlcswitch/interceptable_switch.go:452]: More validation is needed here, eg: max HTLC shouldn't be greater than the channel capacity, etc, etc. I think we should just call the normal validation routine.

**Review by Roasbeef — COMMENTED** [review]
Changes look pretty good, I think the last commit needs to be squashed somewhere, or re-worked slightly. Main question here is re the intended behavior now, and if what is wanted is actually already supported, eg: `lnd` node `N`, makes virtual nodes `M_i...`, then makes invoices with pubkey of node `M`, they can then process the onion twice to decode full contents, and utilize the _existing_ method to send back the error as if it came from node `M`.

File: lnrpc/routerrpc/router.proto (hunk 4 above)
Comment (joostjager) [reply lnrpc/routerrpc/router.proto:927]: >IIUC though, you actually want to be able to send errors from the "virtual node", which means you need to use the shared secret for that hop for the path finding retry/apply logic to work properly.

That is correct.

But sometimes though the virtual node may want to return an error originating from the preceding real lnd node. An example of this is when there is a routing fee associated with the virtual channel. https://github.com/NYDIG-OSS/lnmux/pull/101 illustrates this. The real node has to encrypt the insufficient fee failure message.

Intercepted htlcs do not pass through the switch in the real lnd instance. This means that any forwarding logic (such as the fee check) needs to be replicated by the external interceptor, and the interceptor requires the ability to instruct the real node to produce certain failure messages. 

This has been possible in master for some time (https://github.com/lightningnetwork/lnd/blob/bf5aab9d5246daede3f499af880a5d7be6c6ee88/lnrpc/routerrpc/router.proto#L930) using the `failure_code` field. This however is not a complete API. Ideally the complete failure message structure would need to be replicated on the proto level to be able to signal any failure. This PR takes the approach of accepting an encoded but not yet encrypted failure message to avoid that.

In summary this PR:
* Completes the interceptor API by allowing users to supply a full failure message rather than a restrictive set of failure codes without additional data fields.
* Enable use cases that require different failure messages to be returned, such as a non-zero virtual channel fee.

File: lnrpc/routerrpc/forward_interceptor.go (hunk 18 above)
Comment (joostjager) [reply lnrpc/routerrpc/forward_interceptor.go:131]: I've been trying out multiple ways to make it a pure function, but after the changes in the later commit that extend this function it never really worked out in a clean way. So settled for this.

File: htlcswitch/interceptable_switch.go (hunk 9 above)
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:90]: If node N is the virtual node, then it is the job of node N-1 to return a channel update. The example is again `fee_insufficient`.

Real node N-1 doesn't know about this channel at all (it is virtual), so what happens is that the virtual node N returns a channel update for the virtual channel to N-1, and N-1 just signs the update.

File: htlcswitch/interceptable_switch.go (hunk 19 above)
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:420]: No, because the real node isn't aware of this virtual channel at all. It's all up to the external interceptor logic to keep track of the virtual channel properties. As mentioned in https://github.com/lightningnetwork/lnd/pull/7067#discussion_r1286817930, the external interceptor is taking over part of the responsibilities of the switch.

File: lnrpc/routerrpc/forward_interceptor.go [hunk 21]
```diff
@@ -134,7 +135,7 @@ func (r *forwardInterceptor) resolveFromClient(
 			// message size + hmac + two uint16 lengths. See BOLT
 			// #4.
 			if len(in.FailureMessage) !=
-				lnwire.FailureMessageLength+32+2+2 {
+				lnwire.FailureMessageLength+sha256.Size+2+2 {
```
[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/routerrpc/forward_interceptor.go]: I don't remember the exact history, but may be the case that a later refactor left just this.

File: htlcswitch/interceptable_switch.go (hunk 20 above)
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:452]: Wanted to call

https://github.com/lightningnetwork/lnd/blob/bf5aab9d5246daede3f499af880a5d7be6c6ee88/routing/ann_validation.go#L165

But that would create a circular reference. For now, I just copied over the `max_htlc >= min_htlc` check. Capacity checking is not possible, because the real lnd doesn't know what the capacity is. It's probably infinite because the channel is typically virtual.

If desired I can do a bigger refactor and move the validation logic into `htlcswitch`. But not sure if worth it.

**Comment by joostjager** [comment]
> Changes look pretty good, I think the last commit needs to be squashed somewhere, or re-worked slightly. 

Yes, will squash once I am certain we're all on the same page with this PR.

>Main question here is re the intended behavior now, and if what is wanted is actually already supported, eg: `lnd` node `N`, makes virtual nodes `M_i...`, then makes invoices with pubkey of node `M`, they can then process the onion twice to decode full contents, and utilize the _existing_ method to send back the error as if it came from node `M`.

Hopefully my replies in this round have clarified this. I know it is confusing, but the scope of this PR is not to change anything about errors encrypted by M.

**Comment by JssDWt** [comment]
Is this PR ready to be merged?
I'd love to use the functionality in this PR in Breez's htlc interceptor.

**Comment by JssDWt** [comment]
@saubyk Is this PR planned any time soon?

**Comment by saubyk** [comment]
> @saubyk Is this PR planned any time soon?

Hi @JssDWt I've added this pr to 18's backlog. But at this point, uncertain if it will be included, as there is still a lot of work to get through to complete the in-flight work. Will make a determination in another week or so, if it will stay in the scope or moved to a minor release
